bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,135 @@
+![Corda](https://www.corda.net/wp-content/uploads/2016/11/fg005_corda_b.png)
+
+# CordApp distribution service (CDS)
+
+DOCUMENT MANAGEMENT
+---
+
+## Document Control
+
+| Title                | CordApp distribution service                             |
+| -------------------- | ------------------------------------------------------------ |
+| Date                 | 22 Aug 2018                                                |
+| Author               | Ivan Schasny, Mike Hearn |
+| Distribution         | Design Review Board, Product Management, Solutions Engineering, Platform Delivery |
+| Corda target version | OS                                                   |
+
+## HIGH LEVEL DESIGN
+
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute updates to CorDapps to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios where upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/topic/22686107#190), which is recommended for reading prior to the design.
+
+* BNO should be able to notify members about new CorDapp version availability.
+* BNO should be able to revoke a specific version of CorDapp from their BN. In this context revocation means that BNO should be able to notify members that a specific version of a CorDapp should no longer be considered valid, so members can promptly update their nodes.
```

Comment:
Does this apply to both jars?
For the contract jar, it's a bit more tricky as that is hardcoded into existing transactions (via hash).

Reply:
BNO can't forcefully revoke any jar. He can just try to encourage people to update by letting them know which jars are potentially malicious.

Currently we left the revocation out-of-scope of the first implementation. Btw this is the PR if you are interested: https://github.com/corda/corda-solutions/pull/41

Reply:
Hey @ivanr3, I tried to look at it but it's a bit overwhelming. 

I'm just on a general crusade against the term "CorDapp" because it means 2 different things, and it's confusing.

Reply:
Agree. That is why we tried to keep the distribution aside from installation to not to open a can of worms :)